component/src/main/ballerina/ballerina/net/reliable/httpr/httpr.bal

```diff
@@ -0,0 +1,373 @@
+        blob storableStream = convertToBlob(req, connectorOptions, serviceUri, path, HTTPVerb) ;
+        return storeMessage(storableStream, generateMBConfiguration());
+    }
+}
+
+function storeMessage(blob storableStream, processor:GuaranteedProcessor guaranteedProcessor)(http:Response, http:HttpConnectorError) {
+
+    http:Response response = {};
+    string destinationName = guaranteedProcessor.destinationName;
+
+    // store the message
+    try {
+        transaction with retries(0) {
+            guaranteedProcessor.store(guaranteedProcessor.config, destinationName, storableStream);
+            log:printDebug("message successfully stored in persistent store");
+            response.setJsonPayload({status:"Message successfully received into Ballerina"});
+            response.setStatusCode(202);
+        }
+    } catch (error r) {
+        log:printError("error when storing the message. " + r.msg);
+        response.setJsonPayload({status:"Message failed to receive into Ballerina"});
+        response.setStatusCode(503);
+    }
+    return response, null;
+}
+
+function convertToBlob(http:Request request, http:Options options, string serviceUri, string path, string httpMethod) (blob) {
+    string encodedPayload = util:base64Encode(request.getStringPayload());
+    HttpEPInvoke epInvoke = {request:request, serviceUrl:serviceUri, path:path, httpMethod:httpMethod, connectorOptions:options, payload:encodedPayload};
+    var epInvokeJson,_ = <json> epInvoke;
+    string epInvokeString = epInvokeJson.toString();
+    blob serialized = epInvokeString.toBlob("UTF-8");
+    return serialized;
+}
+
+public function handle (blob objectStream) (error) {
+    endpoint<http:HttpClient> httpEp {
+    }
+
+    // build EPInvoke instance
+    string epInvokeString = objectStream.toString("UTF-8");
+    var epInvokeJson, ex = <json> epInvokeString;
+    var epInvoke, er = <HttpEPInvoke> epInvokeJson;
+
+    // access the request inside EPInvoke instance
+    var request, _ = (http:Request) epInvoke.request;
+
+    // Re-generate request headers from serialized request
+    // Request headers are not getting de-serialized. Therefore we have to do that manually
+    // All the maps are getting de-serialized as <string, any>. if the original map contained <string, struct>
+    map requestHeaders = request.headers;
+    string[] keys = requestHeaders.keys();
+
+    request.removeAllHeaders();
+
+    // copying headers
+    int i = 0;
+    // loop through all the distinct header values in the message
+    while (i < lengthof keys) {
+        // Headers are contained as key,value pairs, but the key can be duplicated. Therefore for each key there is an
+        // array of values. When taking a single key we have to process the whole array of values.
+        // Each value will be a struct with string value and a map (map of string key, string value pairs to contain parameters)
+        var header,_ = (json)requestHeaders[keys[i]];
```

Comment:
Better to catch the returning error (instead of _ ) and rethrow a meaningful error. Because If someone has tampered the serialized json and altered it somehow, there is a possibility of this conversion to fail, and the 'header' will be null. So it can cause null pointers in the downstream code.

Applicable to all places.

Reply:
if there is an error in the json it should be caught at the top level right? Being null is a valid case, i will put a check for that

Reply:
included null checks, and checked type-conversion errors at the top level